staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -569,6 +571,83 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	return nil
 }
 
+// EventHandlerRemovable is an extension interface
+// for a SharedIndexInformer that allows to remove event handlers again.
+// It is possible to remove only event handlers that are (go) comparable,
+// This means event handlers like ResourceEventHandlerFuncs cannot be removed
+// if added by value and not by reference, because they contain function pointers
+// that are not comparable in go. Trying to remove an uncomparable handler
+// results in an appropriate error.
+// This feature can be used to dynamically add and remove watches again depending
+// on the actual state of a controller.
+type EventHandlerRemovable interface {
+	RemoveEventHandler(handler ResourceEventHandler) error
+}
+
+// ExtendedSharedIndexInformer is an extended interface to the
+// SharedIndexInformer informer interface implemented by the standard
+// sharedIndexInformer implementation,
+// It offers the removing of handlers and some state info useful
+// to use after handler removal to decide whether an informer
+// is still required.
+type ExtendedSharedIndexInformer interface {
+	SharedIndexInformer
+	EventHandlerRemovable
+	HasEventHandlers() bool
+	IsStopped() bool
```

Comment:
It gives you more control managing informers besides factories, once you have bare informers in your hands. and you want to explicitly create/delete watches and/or add/remove handlers again.

Reply:
makes sense

Reply:
The creation and deletion of watches is not part of the PR, it could completely be done on top. The scenario for the usage of such a feature are controllers that are controlled by other resources. Just an example, you want to provide indices of arbitrary kubernetes resources, based on a dedicated CRD. Once an index is created for a resource, a watch has to be established (or if already existent, only an additional handler might be configured). If the index is deleted again, the handler and, if the watch is not used by other controllers in the same controller manager, the watch should be removed again. Similar actions would be required, if a controller is potentially responsible for multiple clusters, depending on certain conditions provided by other resources.

The deletion of watches is possible just by a dedicated cancel context (on top) used to initiate the watch and the informer. This works fine (beside the problem described in  #98653) and is minimal invasive. So, all this is potentially possible just by adding the feature of being able to remove handlers again.

If a traditional InformerFactory is used this is for sure not possible without extending the factory, too. But this PR is the precondition to propagate such a feature up the abstraction levels and focuses on the informer itself.

Reply:
thanks for the explanations. i'm not sure i fully understand the use case, two quick questions:
1. do you still need a separate watch if informer is used? put it another way, wouldn't an informer for a resource sufficient, hence no need to use both an informer and a watch on the same resource?
2. since this PR is about removing event handlers, probably it's better to have a separate PR for supporting watches creation/deletion if needed. what do you think?